lib/js/walker.js

```diff
@@ -36,243 +39,59 @@ function ast_walker(){
         scope = array.scope;
 
       for (var i = 0, len = array.length; i < len; i++)
-        walk(array, i);
+        walk(array[i]);
 
       scope = storedScope;
     }
-  };
-
-  function walk(ast, idx){
-    var indexed = typeof idx == 'number';
-    var token = indexed && ast ? ast[idx] : ast;
-
-    if (!token)
-      return ast;
-
-    var storedScope = scope;
-    if (token.scope)
-      scope = token.scope;
-
-    stack.push(token);
-
-    var userFn = user[token[0]] || user['*'];
-    if (userFn)
-    {
-      var ret = userFn.call(walkerContext, token, scope);
-
-      if (ret != null)
-      {
-        if (indexed && Array.isArray(ret))
-          ast[idx] = ret;
-
-        scope = storedScope;
-        stack.pop();
-
-        return ret;
-      }
-
-      if (token.scope)
-        scope = token.scope;
-    }
-
-
-    switch (token[0])
-    {
-      case 'toplevel': // type, statements
-      case 'block':    // type, statements
-      case 'splice':   // type, statements
-      case 'array':    // type, elements
-
-        //if (token[1]) walkEach(token[1]); // statements or elements
-
-        var body = token[1];
-        if (body)
-          for (var i = 0, len = body.length; i < len; i++)
-            walk(body, i);
-
-        break;
-
-      case 'seq':      // type, ...tokens
-
-        for (var i = 1, len = token.length; i < len; i++)
-          walk(token, i);
-
-        break;
-
-      case 'var':      // type, defs
-      case 'const':    // type, defs
-
-        for (var i = 0, defs = token[1], def; def = defs[i]; i++)
-          if (def[1]) // value; optional for var defs
-            walk(def, 1);
+  }
 
-        break;
+  function walk(ast){
+    estraverse.traverse(ast, {
+      enter: function(token){
+        tokenStack.push(token);
+        if (token.scope){
+          scope = token.scope;
+          scopeStack.push(scope)
+        }
 
-      case 'object':   // type, props
+        var userFn = user[token.type] || user['*'];
 
-        for (var i = 0, props = token[1], prop; prop = props[i]; i++)
-          walk(prop, 1);
+        if (userFn)
+        {
+          var managerContext = Object.create(walkerContext);
```

Comment:
I suppose those 3 lines aren't performant. You create a new object and two bounded functions for _each_ node on walking.